src/main/java/bio/terra/workspace/service/resource/controlled/mappings/ControlledResourceInheritanceMapping.java

```diff
@@ -0,0 +1,91 @@
+package bio.terra.workspace.service.resource.controlled.mappings;
+
+import bio.terra.workspace.service.iam.model.ControlledResourceIamRole;
+import bio.terra.workspace.service.iam.model.WsmIamRole;
+import bio.terra.workspace.service.resource.controlled.AccessScopeType;
+import bio.terra.workspace.service.resource.controlled.ManagedByType;
+import bio.terra.workspace.service.workspace.exceptions.InternalLogicException;
+import com.google.common.collect.ImmutableMultimap;
+import com.google.common.collect.Multimap;
+
+/**
+ * Static mappings of workspace-level IAM roles to resource-level IAM roles.
+ *
+ * <p>This mapping represents which resource-level roles are granted to workspace members based on
+ * their workspace-level role. This mapping is slightly different for each category of controlled
+ * resource (user-shared, user-private, app-shared, app-private), so this class holds four separate
```

Comment:
Yeah - this is something we went back and forth on earlier on with controlled resources. We ultimately decided against it because it would only be used in a few spots for IAM, but I still use that notion of a category a lot while describing things.